### PR TITLE
Fix group filter for ldap source sync

### PR DESCRIPTION
**Pull request by pezhovski**
There are 2 separate flows of creating a user: authentication and source sync.
When a group filter is defined, source sync ignores group filter, while authentication respects it.
With this PR I've fixed this behavior, so both flows now apply this filter when searching users in LDAP in a unified way.

- Unified LDAP group membership lookup for authentication and source sync flows
- Replaced custom group membership lookup (used for authentication flow) with an existing listLdapGroupMemberships method (used for source sync flow)
- Modified listLdapGroupMemberships and getUserAttributeListedInGroup in a way group lookup could be called separately
- Added user filtering based on a group membership for a source sync
- Added tests to cover this logic

### CODE REVIEW

File: services/auth/source/ldap/source_search.go [hunk 1]
```diff
@@ -196,22 +196,39 @@ func checkRestricted(l *ldap.Conn, ls *Source, userDN string) bool {
 }
 
 // List all group memberships of a user
-func (source *Source) listLdapGroupMemberships(l *ldap.Conn, uid string) []string {
+func (source *Source) listLdapGroupMemberships(l *ldap.Conn, uid string, aplyGroupFilter bool) []string {
```
Comment (silverwind) [inline services/auth/source/ldap/source_search.go]:
```suggestion
func (source *Source) listLdapGroupMemberships(l *ldap.Conn, uid string, applyGroupFilter bool) []string {
```

File: services/auth/source/ldap/source_search.go [hunk 2]
```diff
@@ -196,22 +196,39 @@ func checkRestricted(l *ldap.Conn, ls *Source, userDN string) bool {
...
+	if !ok {
+		return ldapGroups
+	}
+
+	groupDN, ok := source.sanitizedGroupDN(source.GroupDN)
+	if !ok {
+		return ldapGroups
+	}
+
+	if aplyGroupFilter {
```
Comment (silverwind) [inline services/auth/source/ldap/source_search.go]:
```suggestion
	if applyGroupFilter {
```

File: services/auth/source/ldap/source_search.go [hunk 3]
```diff
@@ -260,6 +275,14 @@ func (source *Source) getMappedMemberships(l *ldap.Conn, uid string) (map[string
 	return membershipsToAdd, membershipsToRemove
 }
 
+func (source *Source) getUserAttributeListedInGroup(entry *ldap.Entry) string {
+	if source.UserUID == "dn" || source.UserUID == "DN" {
```
Comment (silverwind) [inline services/auth/source/ldap/source_search.go]:
```suggestion
	if strings.ToLower(source.UserUID) == "dn" {
```

File: services/auth/source/ldap/source_search.go (hunk 1 above)
Comment (pezhovski) [reply services/auth/source/ldap/source_search.go]: Thanks, fixed

File: services/auth/source/ldap/source_search.go (hunk 3 above)
Comment (pezhovski) [reply services/auth/source/ldap/source_search.go]: Fixed

**Review by silverwind — APPROVED** [review]
Seems reasonable, thought I'm not an expert around these parts.

**Review by zeripath — APPROVED** [review]

**Comment by lunny** [comment]
make LG-TM work

**Comment by KN4CK3R** [comment]
I had some headaches merging this with #21441 because the same code was changed. @CrazyTushkan Could you please test with my PR to see if your changes are still working as intended? The test does not fail but I don't know if that covers every aspect.

**Comment by CrazyTushkan** [comment]
@KN4CK3R sorry for the late response, If you still need this to be double checked, I would gladly do it later this week.

**Comment by KN4CK3R** [comment]
No, thank you. The other PR got tested and is already merged.

**Pull request merged**